From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: MrPowerGamerBR <git@mrpowergamerbr.com>
Date: Mon, 6 Jul 2026 13:42:13 -0300
Subject: [PATCH] Assign unique ID for each connection

It is useful when you need to "bootstrap" a connection, that is, you want to query player data from a external service before the player actually logs in

Using the player's unique ID is not a good solution because of race conditions where a player may have multiple connections in flight, so using a unique ID per connection is more safe

diff --git a/net/minecraft/network/Connection.java b/net/minecraft/network/Connection.java
index 76596e87945f0b95af8032338a64817a8faed4d0..bfb31a4e6ec6eac47ef8bd204e77e0da280ba7c1 100644
--- a/net/minecraft/network/Connection.java
+++ b/net/minecraft/network/Connection.java
@@ -65,6 +65,10 @@ public class Connection extends SimpleChannelInboundHandler<Packet<?>> {
     private final PacketFlow receiving;
     private volatile boolean sendLoginDisconnect = true;
     private final Queue<WrappedConsumer> pendingActions = Queues.newConcurrentLinkedQueue(); // Paper - Optimize network
+    // SparklyPaper start - assign unique ID for each connection
+    public static final java.util.concurrent.atomic.AtomicLong nextConnectionId = new java.util.concurrent.atomic.AtomicLong(0);
+    public long connectionId = nextConnectionId.getAndIncrement();
+    // SparklyPaper end
     public Channel channel;
     public SocketAddress address;
     // Spigot start
@@ -892,7 +896,7 @@ public class Connection extends SimpleChannelInboundHandler<Packet<?>> {
                     /* Player was logged in, either game listener or configuration listener */
                     final com.mojang.authlib.GameProfile profile = commonPacketListener.getOwner();
                     new com.destroystokyo.paper.event.player.PlayerConnectionCloseEvent(profile.id(),
-                        profile.name(), this.address, false).callEvent();
+                        profile.name(), this.address, false, this.connectionId).callEvent(); // SparklyPaper - assign unique ID for each connection
                 } else if (packetListener instanceof net.minecraft.server.network.ServerLoginPacketListenerImpl loginListener) {
                     /* Player is login stage */
                     switch (loginListener.state) {
@@ -902,7 +906,7 @@ public class Connection extends SimpleChannelInboundHandler<Packet<?>> {
                         case ACCEPTED:
                             final com.mojang.authlib.GameProfile profile = loginListener.authenticatedProfile; /* Should be non-null at this stage */
                             new com.destroystokyo.paper.event.player.PlayerConnectionCloseEvent(profile.id(), profile.name(),
-                                this.address, false).callEvent();
+                                this.address, false, this.connectionId).callEvent(); // SparklyPaper - assign unique ID for each connection
                     }
                 }
                 // Paper end - Add PlayerConnectionCloseEvent
